// _egg-ts-helper@1.20.0@egg-ts-helper/dist/cmd/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const enquirer_1 = require("enquirer");
const utils = tslib_1.__importStar(require("../utils"));
const path_1 = tslib_1.__importDefault(require("path"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const __1 = require("../");
const TYPE_TS = 'typescript';
const TYPE_JS = 'javascript';
const JS_CONFIG = {
    include: ['**/*'],
    exclude: [
        'node_modules/',
        'app/web/',
        'app/view/',
        'public/',
        'app/mocks/',
        'coverage/',
        'logs/',
    ],
};
const TS_CONFIG = {
    compilerOptions: {
        target: 'es2017',
        module: 'commonjs',
        strict: true,
        noImplicitAny: false,
        experimentalDecorators: true,
        emitDecoratorMetadata: true,
        allowSyntheticDefaultImports: true,
        charset: 'utf8',
        allowJs: false,
        pretty: true,
        lib: ['es6'],
        noEmitOnError: false,
        noUnusedLocals: true,
        noUnusedParameters: true,
        allowUnreachableCode: false,
        allowUnusedLabels: false,
        strictPropertyInitialization: false,
        noFallthroughCasesInSwitch: true,
        skipLibCheck: true,
        skipDefaultLibCheck: true,
        inlineSourceMap: true,
    },
};
class InitCommand {
    constructor() {
        this.description = 'Init egg-ts-helper in your existing project';
        this.options = '<type>';
    }
    async run(_, { args, cwd }) {
        let type = args[1];
        const pkgInfo = utils.getPkgInfo(cwd);
        const jsconfigPath = path_1.default.resolve(cwd, './jsconfig.json');
        const jsConfigExist = fs_1.default.existsSync(jsconfigPath);
        const typeList = [TYPE_TS, TYPE_JS];
        pkgInfo.egg = pkgInfo.egg || {};
        // verify type
        if (!typeList.includes(type)) {
            const result = await enquirer_1.prompt({
                type: 'autocomplete',
                name: 'type',
                message: 'Choose the type of your project',
                choices: jsConfigExist ? typeList.reverse() : typeList,
            }).catch(() => {
                utils.log('cancel initialization');
                return { type: '' };
            });
            type = result.type;
        }
        if (type === TYPE_JS) {
            // create jsconfig.json
            if (!jsConfigExist) {
                utils.log('create ' + jsconfigPath);
                fs_1.default.writeFileSync(jsconfigPath, JSON.stringify(JS_CONFIG, null, 2));
            }
        }
        else if (type === TYPE_TS) {
            pkgInfo.egg.typescript = true;
            // create tsconfig.json
            const tsconfigPath = path_1.default.resolve(cwd, './tsconfig.json');
            if (!fs_1.default.existsSync(tsconfigPath)) {
                utils.log('create ' + tsconfigPath);
                fs_1.default.writeFileSync(tsconfigPath, JSON.stringify(TS_CONFIG, null, 2));
            }
        }
        else {
            return;
        }
        // add egg-ts-helper/register to egg.require
        pkgInfo.egg.require = pkgInfo.egg.require || [];
        if (!pkgInfo.egg.require.includes('egg-ts-helper/register')) {
            pkgInfo.egg.require.push('egg-ts-helper/register');
        }
        // write package.json
        const pkgDist = path_1.default.resolve(cwd, './package.json');
        fs_1.default.writeFileSync(pkgDist, JSON.stringify(pkgInfo, null, 2));
        utils.log('change ' + pkgDist);
        // build once
        utils.log('create d.ts ...');
        __1.createTsHelperInstance({ cwd }).build();
        utils.log('complete initialization');
    }
}
exports.default = new InitCommand();
